src/store/index.js

```diff
@@ -1,14 +1,33 @@
 import { createStore } from 'redux';
 import reducer from '../reducers';
-import { saveState } from '../utils';
+import { resetData } from '../actions';
+import getInitialState from './initial-state';
+import loadData from './load-data';
+import { saveState } from './helpers';
+import checkFontLoaded from './check-font-loaded';
+
+export default function configureStore(props) {
+  /**
+   * Dispatch an action to update the store with new pipeline data
+   * @param {Object} normalizedData Normalised state data
+   */
+  const resetStoreData = normalizedData => {
+    if (store) {
+      store.dispatch(resetData(normalizedData));
+    }
+  };
+
+  const pipelineData = loadData(props.data, resetStoreData);
+  const initialState = getInitialState(pipelineData, props);
 
-export default function configureStore(initialState) {
   const store = createStore(reducer, initialState);
 
   store.subscribe(() => {
     const { textLabels, theme, typeDisabled } = store.getState();
     saveState({ textLabels, theme, typeDisabled });
   });
 
+  checkFontLoaded(store);
+
```

Comment:
This seems like a view level check rather than something that should be in the store (the store doesn't care about fonts rather the application state) - I'd recommend moving to componentDidMount on App as a place for this check

Reply:
Can do - I was just hoping to keep this stuff separated from App and avoid having App do too much. To me it seems more store-related, as it's not directly affecting rendering 🤷‍♂

Reply:
## Description

It's not related to store (it only manipulates the state through actions), it's an action firing to do some async check.

Looking at checkFontLoaded it is in fact a "thunk" i.e. an action that can have side effects (e.g. triggering other actions, async things etc) Example from redux documentation:

## Redux thunk example

https://redux.js.org/advanced/async-actions/

```
export const REQUEST_POSTS = 'REQUEST_POSTS'
function requestPosts(subreddit) {
  return {
    type: REQUEST_POSTS,
    subreddit
  }
}
export const RECEIVE_POSTS = 'RECEIVE_POSTS'
function receivePosts(subreddit, json) {
  return {
    type: RECEIVE_POSTS,
    subreddit,
    posts: json.data.children.map(child => child.data),
    receivedAt: Date.now()
  }
}
export const INVALIDATE_SUBREDDIT = 'INVALIDATE_SUBREDDIT'
export function invalidateSubreddit(subreddit) {
  return {
    type: INVALIDATE_SUBREDDIT,
    subreddit
  }
}
// Meet our first thunk action creator!
// Though its insides are different, you would use it just like any other action creator:
// store.dispatch(fetchPosts('reactjs'))
export function fetchPosts(subreddit) {
  // Thunk middleware knows how to handle functions.
  // It passes the dispatch method as an argument to the function,
  // thus making it able to dispatch actions itself.
  return function(dispatch) {
    // First dispatch: the app state is updated to inform
    // that the API call is starting.
    dispatch(requestPosts(subreddit))
    // The function called by the thunk middleware can return a value,
    // that is passed on as the return value of the dispatch method.
    // In this case, we return a promise to wait for.
    // This is not required by thunk middleware, but it is convenient for us.
    return fetch(`https://www.reddit.com/r/${subreddit}.json`)
      .then(
        response => response.json(),
        // Do not use catch, because that will also catch
        // any errors in the dispatch and resulting render,
        // causing a loop of 'Unexpected batch number' errors.
        // https://github.com/facebook/react/issues/6895
        error => console.log('An error occurred.', error)
      )
      .then(json =>
        // We can dispatch many times!
        // Here, we update the app state with the results of the API call.
        dispatch(receivePosts(subreddit, json))
      )
  }
}
```

This coupled with introducing redux-thunk:

```
const store = createStore(
  rootReducer,
  applyMiddleware(
    thunkMiddleware,
  )
)
```

## Steps to change here

- Introduce thunkMiddleware
- Update createStore
- Move checkFontLoaded into actions folder (along with tests)
- Rejig it to return a function like the example above
- Call the action in App (e.g. componentDidMount)

Reply:
As discussed on Slack: Good idea but right now it doesn't simplify the code but it does add a bunch of extra complexity, so I'm gonna shelve this for now and revisit it later if we need to add more middleware 👍